Batching blog post (#30)
resonatehq/blog: blog/2024-10-22-transparent-batching-with-the-resonate-python-sdk/index.md

:::tip RSVP for the webinar

[RSVP for the upcoming webinar](https://www.resonatehq.io/webinars) to learn more about the Python SDK!

:::

This post highlights one of the SDK’s features that might make you jump for joy.

Let's talk about batching.

## Problem space

To ensure we’re all on the same page, "batching" refers to the practice of grouping multiple operations into a single unit of work.

**Why do this?**

Because performing operations in batches can be much more efficient than processing them one by one.

When we talk about efficiency we could be talking about two different things:

1. Speed
2. Resource utilization

With speed, the benefit is clear - things happen faster and you are removing bottlenecks.

With resource utilization, the benefit will often come down to cost savings.
In todays cloud environments, you are often charged for resources you use.
If a resource is remote and accessed over a network, you may be charged every time you access it.

This level of efficiency might not be significant at smaller scales, but at larger scales, efficiency becomes crucial.

> For example, processing 100,000 operations sequentially can be far more costly and time-consuming than batching those 100,000 operations into a single task. Of course, the operations must be similar and the resources being used need a way to group the operations together.

Author: ```suggestion
For example, processing 100,000 operations sequentially can be far more costly and time-consuming than batching those 100,000 operations into a single task. Of course, the target system must support batching.
```

So let's focus on a concrete example.

Consider an application that creates a row in a database for each new user.

If the application gets less than 1 new user per minute, or even every 30 seconds, then it can probably handle sequential inserts.
In other words, the application can run and commit one SQL query per new user and there may not ever be a problem.

```python
def _create_user(user: int):
    conn.execute("INSERT INTO users (value) VALUES (?)", (user,))
    conn.commit()
    print(f"User {user} has been inserted to database")
```

But what if the application received 1000 new users per second?

> The database would lock on each query and the application probably wouldn't be able to keep up with all the demand.

Author: ```suggestion
The database would save to durable storage on each query and the application probably wouldn't be able to keep up with all the demand.
```

Each second there would be a longer and longer delay for users to get added to the database.

> We can imagine that, if these user creation requests were synchronous, and coming over a network, that many of them would timeout while waiting for their turn to get a row in the database.

Author: ```suggestion
We can imagine that, if these user creation requests were coming over a network, that many of them would timeout while waiting for their turn to get a row in the database.
```

In this very simplified example, we might now consider batching SQL queries so that more user rows are created per commit.

```python
def _create_users_in_batch(users: list[int]):
   first_user = users[0]
   last_user = None
   for user in users:
       conn.execute("INSERT INTO users (value) VALUES (?)", (user,))
       last_value = user
   conn.commit()
   print(f"Users from {first_user} to {last_user} have been inserted to database.")
```

If we can batch thousands of queries into a single commit, then likely the application would be able to keep up with the demand.

> :::tip Do not use in production
> 
> In production, you would want to ensure that the new user writes are idempotent and either check to see if the row exists already or update if already does exist.

Author: ```suggestion
:::tip In Production

In production, ensure that inserts are idempotent to account for the possibility of retries.
```

def _batch_handler(_: Context, users: list[InsertUser]):
    first_value = users[0].id
    last_value = None
    for user in users:
        conn.execute("INSERT INTO users (value) VALUES (?)", (user.id,))
        last_value = user.id
    conn.commit()
    print(f"Values from {first_value} to {last_value} have been inserted to database.")
```

<!--SNIPEND-->

Next, register the data structure and the handler with the Resonate Scheduler.

<!--SNIPSTART examples-py-features-batching-init {"selectedLines":["3-5","13-14", "40-41"]}-->

[features/batching/src/batching/**init**.py](https://github.com/resonatehq/examples-py/blob/main/features/batching/src/batching/__init__.py)

```py
# ...
from resonate.scheduler import Scheduler
from resonate.storage import LocalPromiseStore
from resonate.retry_policy import never
# ...
# Create a Resonate Scheduler with an in memore promise store
resonate = Scheduler(LocalPromiseStore(), processor_threads=1)
# ...
# Register the batch handler and data structure with the Resonate Scheduler
resonate.register_command_handler(InsertUser, _batch_handler, retry_policy=never())
```

<!--SNIPEND-->

> Finally, create a top-level function that can be invoked over and over again and passes the data to Resonate to manage.

Author: ```suggestion
Finally, create a function that can be invoked over and over again and passes the data to Resonate to manage.
```
